qiskit/algorithms/optimizers/gradient_descent.py

```diff
@@ -1,136 +1,47 @@
-    gradient.
-
-    Examples:
-
-        A minimum example that will use finite difference gradients with a default perturbation
-        of 0.01 and a default learning rate of 0.01.
-
-        .. code-block::python
-
-            from qiskit.algorithms.optimizers import GradientDescent
-
-            def f(x):
-                return (np.linalg.norm(x) - 1) ** 2
-
-            initial_point = np.array([1, 0.5, -0.2])
-
-            optimizer = GradientDescent(maxiter=100)
-            x_opt, fx_opt, nfevs = optimizer.optimize(initial_point.size,
-                                                      f,
-                                                      initial_point=initial_point)
-
-            print(f"Found minimum {x_opt} at a value of {fx_opt} using {nfevs} evaluations.")
+from qiskit.algorithms.algorithm_result import AlgorithmResult
 
-        An example where the learning rate is an iterator and we supply the analytic gradient.
-        Note how much faster this convergences (i.e. less ``nfevs``) compared to the previous
-        example.
 
-        .. code-block::python
-
-            from qiskit.algorithms.optimizers import GradientDescent
-
-            def learning_rate():
-                power = 0.6
-                constant_coeff = 0.1
-
-                def powerlaw():
-                    n = 0
-                    while True:
-                        yield constant_coeff * (n ** power)
-                        n += 1
+from .optimizer import Optimizer, OptimizerSupportLevel, OptimizerResult, POINT
+from .steppable_optimizer import AskObject, TellObject, OptimizerState, SteppableOptimizer, CALLBACK
 
-                return powerlaw()
 
-            def f(x):
-                return (np.linalg.norm(x) - 1) ** 2
+def constant(eta=0.01):
+    """Yield a constant."""
 
-            def grad_f(x):
-                return 2 * (np.linalg.norm(x) - 1) * x / np.linalg.norm(x)
+    while True:
+        yield eta
 
-            initial_point = np.array([1, 0.5, -0.2])
 
-            optimizer = GradientDescent(maxiter=100, learning_rate=learning_rate)
-            x_opt, fx_opt, nfevs = optimizer.optimize(initial_point.size,
-                                                      f,
-                                                      gradient_function=grad_f,
-                                                      initial_point=initial_point)
+@dataclass
+class GD_OptimizerState(OptimizerState):
 
-            print(f"Found minimum {x_opt} at a value of {fx_opt} using {nfevs} evaluations.")
+    eta: Union[float, Callable[[], Iterator]]
+    stepsize: float
 
-    """
 
+class GradientDescent(SteppableOptimizer):
     def __init__(
         self,
-        maxiter: int = 100,
+        maxiter: int = 1000,
+        callback: Optional[CALLBACK] = None,
         learning_rate: Union[float, Callable[[], Iterator]] = 0.01,
         tol: float = 1e-7,
-        callback: Optional[CALLBACK] = None,
```

Comment:
Could we keep the original default values and order of the arguments? 🙂

Reply:
Yes, this is actually a very good point. I absolutely need to make sure that we don't make any old code obsolete by some kind of mistake of this kind.